### ISSUE TITLE
Redux actions not working/not triggered when app is in background on iOS

### ISSUE THREAD

**Issue by LEEY19** [OneSignal/react-native-onesignal#264]
So i have something like this:
```
  async onReceived(notification) {
    store.dispatch(addNotif({
      title: notification.payload.title,
      description: notification.payload.body,
      object: JSON.parse(notification.payload.additionalData.object),
      read: false,
    }));
  }
```
I found out that whatever that is in onReceived is not run when the app is in background on iOS when debugger is not running on xcode. It works fine on Android.

In another piece of code:
```
  async onOpened(openResult) {
    store.dispatch(toggleRead(openResult.notification.payload.body));
    if ((openResult.notification.payload.additionalData.type === "Post") || (openResult.notification.payload.additionalData.type === "Comment")) {
      let object = await JSON.parse(openResult.notification.payload.additionalData.object);
      this.refs.root_navigator.push({ index: 'ViewPost', post: object.post});
    } else {
      let object = await JSON.parse(openResult.notification.payload.additionalData.object);
      this.refs.root_navigator.push({ index: 'ViewPost', post: object.post, comment: object.comment});
    }
  }
```
The store.dispatch in onOpened is not run when notification is tapped, but whatever that comes after (the if logic etc) is processed. But then again, it works fine on Android. I am not sure if its a redux issue or an issue with this library. Appreciate if you could look into this.

**Comment by LEEY19** [comment]
@avishayil just validated, store.dispatch in onOpened actually runs. It is the store.dispatch in onReceived that is not run. So now i don't think its a redux issue, its issue specifically to onReceived function in iOS

**Comment by zakster12** [comment]
For me onReceived works only when the user is in the app (the app is active) or in the background but not when the app is closed. I receive them always in the notification center (iOS) but not in my app's:
```
 onReceived = (receivedNotif) => {
     ...
}
```
So I am also looking into this: How we can receive notifications in the app when the app is **closed** or the user doesn't press on the notification?

**Comment by LEEY19** [comment]
i have updates - to activate onReceived when app is on **background** on iOS, you have to enable Background App Refresh for the app. However, when its completely killed, we might have to look into Pushkit iOS. 

Right now, I am facing the problem of, onReceive not being called when app is **completely killed** for both Android and iOS. Hope someone can shed more light on this. 

@zakster12 lets keep the discussion active since we are exploring the same thing

**Comment by zakster12** [comment]
@LEEY19 I have exactly the same problem:

> Right now, I am facing the problem of, onReceive not being called when app is completely killed for both Android and iOS

From @jkasten2:

> If you need to do background processing in your app as soon as the notification arrives you will need to set content_available to true on our REST API or enable this option on the dashboard when creating a new message. Enabling this does not fire the callback above, it instead wakes up your app and calls the application:didReceiveRemoteNotification:fetchCompletionHandler: selector. This selector isn't available in Cordova so you will need to write Objective-C or Swift code to handle this.

I have content_available set to true and I receive the notifications when the app is in the background but when it's completely killed/closed I don't receive them in the _onReceived_ method as @LEEY19 said. 

@avishayil @jkasten2 any help here?

**Comment by LEEY19** [comment]
@avishayil I have been dabbling with your native codes in RNOneSignal.java, as I am trying to implement WakefulBroadcastReceiver (referring to https://stackoverflow.com/questions/24313539/push-notifications-when-app-is-closed) so that the device can be woken app when it is originally **completely killed**. I realised that you are just using the normal BroadcastReceiver. However, because I don't code in java, its extremely hard for me to modify your native codes. Your help will be valuable.

**Comment by zakster12** [comment]
We have the same problem on iOS (I am testing only on iOS).

**Comment by kiroskirin** [comment]
I'm not sure this will help.
When run `react-native link react-native-onesignal`, 

it will create `Header Search Paths` like this
`$(SRCROOT)/../node_modules/react-native-onesignal/ios/RCTOneSignal`

just change it to
`$(SRCROOT)/../node_modules/react-native-onesignal/ios`

I tested on iOS device and it work even when app got killed.

**Comment by kesiena115** [comment]
I'm having this problem on Android i.e., onReceived is not running when the app is killed. Anyone found a solution to this?

**Comment by hcyildirim** [comment]
I'm facing the same problem, is there any solution to this issue?

**Comment by therahuls1ngh** [comment]
same issue :(

**Comment by jordanmkoncz** [comment]
Just a heads up that I've posted a pretty lengthy comment at https://github.com/geektimecoil/react-native-onesignal/issues/332#issuecomment-341326624 discussing this problem.

**Comment by wkoutre** [comment]
I've posted a [code sample](https://github.com/geektimecoil/react-native-onesignal/issues/332#issuecomment-349856950) resulting in the `opened` event listener being called after tapping a notification from a completely closed AppState.

I know issues here are more focused on the `received` event, aka in the event of the User entering the app without tapping an outstanding notification. There must be a way to check the notification queue, especially considering there's the `OneSignal.clearOneSignalNotifications` method available for Android.

**Comment by Nightsd01** [comment]
@LEEY19 I am closing this issue due to inactivity.

We have recently made some changes in recent releases for how the `onReceived` and `onOpened` events are handled so I think this issue is likely resolved at this point.

If you (or anyone else) is still encountering this issue please respond and I will be happy to reopen and investigate the problem.

**Comment by isaaclem** [comment]
@Nightsd01 : Nope it's not solved, I'm still facing it at version `^3.2.6`, whereby on iOS platform, the event registered wont be fired onReceived if apps is on background

**Comment by harutyundr** [comment]
Still having this issue on using "react-native-onesignal": "^3.4.1" 

When the app is in background it does receive the notification while running a debug build via XCode, but the Release build tested via Testflight does not receive it. 

However, during the I noticed that maybe one-two times the callback was actually called even with the app in the background. So I guess there is some management by iOS which sometimes does activate the app sometimes does not. 

I am sending both mutable_content and content_available from the back-end.
